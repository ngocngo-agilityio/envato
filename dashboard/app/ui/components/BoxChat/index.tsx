'use client';

import { memo, useCallback, useEffect, useRef, useState } from 'react';
import { Box, Heading, Flex } from '@chakra-ui/react';

// Components
import { InputSendMessages, Loading, Message } from '@/ui/components';

// Interface
import { TMessages } from '@/lib/interfaces';

// Stores
import { authStore } from '@/lib/stores';

// Hooks
import { getInfoRoomChat, useSubscribeToChat } from '@/lib/hooks';

// Firebase
import { doc, getDoc, setDoc } from 'firebase/firestore';

// Utils
import { db } from '@/lib/utils';

// Constants
import { AUTHENTICATION_ROLE, FIREBASE_CHAT } from '@/lib/constants';

const initialUserChat = {
  roomChatId: '',
  userId: '',
  adminId: '',
  avatarUrl: '',
  avatarAdminUrl: '',
  displayName: '',
};

const BoxChat = () => {
  const [messages, setMessages] = useState<TMessages[]>([]);
  const [isLoading, setIsLoading] = useState(true);
  const user = authStore((state) => state.user);
  const hasPermission = user?.role === AUTHENTICATION_ROLE.MEMBER;
  const [userChat, setUserChat] = useState(initialUserChat);
  const boxRef = useRef<HTMLDivElement | null>(null);

  const { avatarAdminUrl, avatarUrl, adminId } = userChat || {};

  const fetchData = async () => {
    const usersData = await getInfoRoomChat(user);

    // Check if usersData is undefined before accessing its properties
    if (usersData) {
      const res = await getDoc(
        doc(db, FIREBASE_CHAT.CHATS, usersData.roomChatId),
      );

      setUserChat(usersData);
      res.exists() ? setMessages(res.data().messages) : await createChatRoom();
    }
    setIsLoading(false);
  };

  const createChatRoom = useCallback(async () => {
    // Get user data
    const usersData = await getInfoRoomChat(user);

    if (usersData)
      await setDoc(doc(db, FIREBASE_CHAT.CHATS, usersData.roomChatId), {
        messages: [],
      });
  }, [user]);

  useEffect(() => {
    fetchData();
  }, []);

  useSubscribeToChat(userChat.roomChatId, setMessages, boxRef);

  return (
    hasPermission && (
      <Box w="full" bg="background.body.quaternary" borderRadius="lg">
        <Flex
          direction="row"
          justifyContent="space-between"
          alignItems="center"
          justify="center"
          borderBottom="1px solid"
          borderBottomColor="border.tertiary"
          padding="24px 26px"
        >
          <Heading
            as="h3"
            fontWeight="semibold"
            color="text.primary"
            fontSize="2xl"
            textTransform="capitalize"
          >
            team chat
          </Heading>
        </Flex>

        {isLoading ? (
          <Loading />
        ) : (
          <Box padding={{ base: '24px 20px', lg: '45px 35px' }}>
            <Box
              ref={boxRef}
              overflowX="auto"
              overflowY="scroll"
              css={{
                '&::-webkit-scrollbar': {
                  width: 2,
                },
                '&::-webkit-scrollbar-track': {
                  width: 2,
                },
                '&::-webkit-scrollbar-thumb': {
                  background: 'gray',
                  borderRadius: '24px',
                },
              }}
              maxHeight={361}
              padding={5}
            >
              {messages.map((message) => {
                const { text = '', date, senderId = '' } = message || {};
                const { seconds = 0 } = date || {};

                return (
                  <Message
                    content={text}
                    key={seconds}
                    senderId={senderId}
                    avatarAdmin={avatarAdminUrl}
                    avatarUser={avatarUrl}
                    superAdminId={adminId}
                  />
                );
              })}
            </Box>

            <InputSendMessages boxRef={boxRef} />
          </Box>
        )}
      </Box>
    )
  );
};

const BoxChatMemorized = memo(BoxChat);

export default BoxChatMemorized;
