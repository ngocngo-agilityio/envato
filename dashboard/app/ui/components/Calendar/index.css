/* Month view */
.rbc-month-view {
  border-width: 0;
}

.rbc-month-header .rbc-header {
  font-weight: 600;
  font-size: 20px;
  line-height: 28px;
  color: var(--c-calendar-m-header-title);
  border: none !important;
}

.rbc-month-header {
  padding-top: 20px;
  border-width: 0;
  border-top: 1px solid
    light-dark(
      var(--c--calendar-m-header-border-light),
      var(--c-calendar-m-header-border-dark)
    );
  margin-bottom: 30px;
}

.rbc-month-header .rbc-header:first-child {
  color: var(--c-calendar-m-header-first-title);
}

.rbc-month-row .rbc-date-cell {
  font-size: 20px;
  line-height: 20px;
  color: light-dark(
    var(--c-calendar-m-cell-light),
    var(--c-calendar-m-cell-dark)
  );
  background: transparent;
  border: none !important;
  text-align: center;
}

.rbc-month-row .rbc-date-cell:first-child,
.rbc-date-cell:last-child {
  color: var(--c-calendar-m-first-last-column-cell);
}

.rbc-month-row .rbc-now.rbc-date-cell {
  color: var(--c-calendar-m-today-cell);
}

.rbc-month-row .rbc-now.rbc-date-cell .rbc-button-link {
  font-weight: 700;
}

.rbc-month-row .rbc-off-range.rbc-date-cell {
  color: var(--c-calendar-m-off-range-cell);
}

.rbc-month-row .rbc-off-range.rbc-date-cell .rbc-button-link {
  font-weight: 500;
}

.rbc-month-row .rbc-off-range.rbc-date-cell:first-child,
.rbc-off-range.rbc-date-cell:last-child {
  color: var(--c-calendar-m-first-last-column-off-range-cell);
}

.rbc-month-row .rbc-off-range-bg {
  background-color: transparent;
}

.rbc-month-row .rbc-button-link {
  font-weight: 500;
}

.rbc-month-row .rbc-button-link:hover {
  text-decoration: underline;
  text-decoration-color: light-dark(
    var(--c-calendar-m-cell-light),
    var(--c-calendar-m-cell-dark)
  );
}

.rbc-month-row .rbc-off-range .rbc-button-link:hover {
  text-decoration: underline;
  text-decoration-color: var(--c-calendar-m-off-range-cell);
}

.rbc-month-row .rbc-date-cell:first-child .rbc-button-link:hover,
.rbc-month-row .rbc-date-cell:last-child .rbc-button-link:hover {
  text-decoration: underline;
  text-decoration-color: var(--c-calendar-m-first-last-column-cell);
}

.rbc-month-row .rbc-off-range:first-child .rbc-button-link:hover,
.rbc-month-row .rbc-off-range:last-child .rbc-button-link:hover {
  text-decoration: underline;
  text-decoration-color: var(--c-calendar-m-first-last-column-off-range-cell);
}

.rbc-month-row .rbc-today {
  background-color: transparent;
}

.rbc-month-row,
.rbc-day-bg {
  border: none !important;
}

.rbc-month-view .rbc-event {
  font-size: 14px;
  line-height: 24px;
  font-weight: 700;
  color: light-dark(
    var(--c-calendar-m-event-light),
    var(--c-calendar-m-event-dark)
  );
  background-color: transparent;
}

.rbc-month-view .rbc-event:hover {
  background-color: light-dark(
    var(--c-calendar-m-event-hover-light),
    var(--c-calendar-m-event-hover-dark)
  );
}

.rbc-month-view .rbc-event-content::before {
  content: '';
  display: inline-block;
  width: 8px;
  height: 8px;
  border-radius: 50%;
  background-color: var(--bg-calendar-m-event-dot);
  margin-right: 6px;
}

/* Time view */
.rbc-events-container .rbc-event {
  background-color: var(--bg-calendar-t-event);
  border-color: var(--c-calendar-t-event-border);
  min-height: 50px !important;
}

.rbc-events-container .rbc-event:focus {
  outline: none;
  box-shadow:
    0 6px 10px 0px
      light-dark(
        var(--box-shadow-t-event-primary-light),
        var(--box-shadow-t-event-primary-dark)
      ),
    0 1px 18px 0px
      light-dark(
        var(--box-shadow-t-event-secondary-light),
        var(--box-shadow-t-event-secondary-dark)
      ),
    0 3px 5px -1px light-dark(var(--box-shadow-t-event-tertiary-light), var(--box-shadow-t-event-tertiary-dark));
}

.rbc-time-view,
.rbc-time-header-content,
.rbc-time-view .rbc-header {
  border: none !important;
}

.rbc-time-header-content .rbc-day-bg {
  border-left: 1px solid var(--c-calendar-t-day) !important;
}

.rbc-time-slot {
  font-size: 14px;
  text-align: right;
}

.rbc-time-header-content .rbc-button-link {
  font-weight: 500;
}

.rbc-time-header-content .rbc-header {
  color: var(--c-calendar-t-header);
}

.rbc-time-header-content .rbc-header:first-child {
  color: var(--c-calendar-t-header-first);
}

.rbc-time-view .rbc-today {
  background-color: light-dark(
    var(--bg-calendar-t-today-light),
    var(--bg-calendar-t-today-dark)
  );
}

.rbc-day-slot,
.rbc-day-bg,
.rbc-date-cell,
.rbc-row-content {
  cursor: pointer;
}

.rbc-time-view .rbc-event .rbc-event-label {
  font-size: 13px;
}

.rbc-time-view .rbc-event .rbc-event-content {
  font-size: 14px;
  font-weight: 500;
  line-height: 16px;
}

/* Calendar */
.rbc-calendar {
  min-height: 530px;
}

@media screen and (min-width: 768px) {
  /* Time view */
  .rbc-time-header-content .rbc-button-link {
    font-size: 16px;
  }

  /* Calendar view */
  .rbc-calendar {
    min-height: 700px;
  }
}

@media screen and (min-width: 992px) {
  /* Time view */
  .rbc-time-header-content .rbc-button-link {
    font-size: 18px;
  }

  /* Calendar view */
  .rbc-calendar {
    min-height: 870px;
  }
}
