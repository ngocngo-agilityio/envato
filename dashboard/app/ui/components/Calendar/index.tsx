'use client';

// Libs
import dynamic from 'next/dynamic';
import { useCallback, useMemo, useState, memo } from 'react';
import { CloseButton, Flex, Heading, useDisclosure } from '@chakra-ui/react';
import { DeleteIcon, EditIcon } from '@chakra-ui/icons';
import {
  Calendar as BigCalendar,
  dayjsLocalizer,
  Views,
  SlotInfo,
  CalendarProps as BigCalendarProps,
  Event,
} from 'react-big-calendar';
import isEqual from 'react-fast-compare';
import dayjs from 'dayjs';

// Components
import {
  CustomToolBar,
  EventForm,
  EventDetails,
  ConfirmDeleteModalBody,
  Indicator,
} from '@/ui/components';

// Types
import { TEvent } from '@/lib/interfaces';

// Constants
import {
  DATE_FORMAT,
  MONTH_DATE_FORMAT,
  TIME_FORMAT_12H,
  TIME_FORMAT_HH_MM,
} from '@/lib/constants';

// Themes
import { useColorfill } from '@/ui/themes/bases/colors';

// Lazy loading components
const Modal = dynamic(() => import('@/ui/components/common/Modal'));

// Styles
import 'react-big-calendar/lib/css/react-big-calendar.css';
import './index.css';

type ViewType = 'month' | 'week' | 'work_week' | 'day' | 'agenda';

interface Slot {
  start: Date;
  end: Date;
}

const localizer = dayjsLocalizer(dayjs);

type CalendarProps = Omit<BigCalendarProps, 'localizer'> & {
  events: (Event & Pick<TEvent, '_id'>)[];
  isLoading?: boolean;
  onAddEvent: (data: Omit<TEvent, '_id'>) => void;
  onEditEvent: (data: TEvent) => void;
  onDeleteEvent: (id: string) => void;
  onSetDate: (date: Date) => void;
};

const Calendar = ({
  events,
  date,
  isLoading = false,
  onAddEvent,
  onEditEvent,
  onDeleteEvent,
  onSetDate,
  ...rest
}: CalendarProps) => {
  const [view, setView] = useState<ViewType>(Views.MONTH);
  const [slot, setSlot] = useState<Slot | null>(null);
  const [selectedEvent, setSelectedEvent] = useState<Event & Partial<TEvent>>();

  const { isOpen: isOpenEventFormModal, onToggle: onToggleEventFormModal } =
    useDisclosure();
  const {
    isOpen: isOpenEventDetailModal,
    onToggle: onToggleEventDetailsModal,
  } = useDisclosure();
  const { isOpen: isOpenConfirmModal, onToggle: onToggleConfirmModal } =
    useDisclosure();
  const { septenary } = useColorfill();

  const {
    _id: selectedEventId = '',
    start: selectedEventStart = '',
    end: selectedEventEnd = '',
    title = '',
  } = selectedEvent || {};
  const selectedEventTitle = title?.toString() || '';

  const { start: startSlot = '', end: endSlot = '' } = slot || {};

  const formattedSelectedEventDate = useMemo(
    () => dayjs(selectedEventStart).format(MONTH_DATE_FORMAT),
    [selectedEventStart],
  );

  const formattedSelectedEventStart = useMemo(
    () => dayjs(selectedEventStart).format(TIME_FORMAT_12H),
    [selectedEventStart],
  );

  const formattedSelectedEventEnd = useMemo(
    () => dayjs(selectedEventEnd).format(TIME_FORMAT_12H),
    [selectedEventEnd],
  );

  const selectedEventTime = useMemo(
    () =>
      `${formattedSelectedEventDate} ${formattedSelectedEventStart} â€“ ${formattedSelectedEventEnd}`,
    [
      formattedSelectedEventDate,
      formattedSelectedEventEnd,
      formattedSelectedEventStart,
    ],
  );

  const startEvent = useMemo(
    () => startSlot || selectedEventStart,
    [selectedEventStart, startSlot],
  );

  const endEvent = useMemo(
    () => endSlot || selectedEventEnd,
    [endSlot, selectedEventEnd],
  );

  const eventDate = useMemo(
    () => dayjs(startEvent).format(DATE_FORMAT),
    [startEvent],
  );

  const eventStartTime = useMemo(
    () => dayjs(startEvent).format(TIME_FORMAT_HH_MM),
    [startEvent],
  );

  const eventEndTime = useMemo(
    () => dayjs(endEvent).format(TIME_FORMAT_HH_MM),
    [endEvent],
  );

  const handleSelectSlot = useCallback(
    (slotInfo: SlotInfo) => {
      setSlot((prev) => ({
        ...prev,
        start: slotInfo.start,
        end: slotInfo.end,
      }));

      // Open the Add event form modal
      onToggleEventFormModal();
    },
    [onToggleEventFormModal],
  );

  const handleOpenEventFormModal = useCallback(() => {
    onToggleEventDetailsModal();
    onToggleEventFormModal();
  }, [onToggleEventDetailsModal, onToggleEventFormModal]);

  const handleAddEvent = useCallback(
    (data: Omit<TEvent, '_id'>) => {
      onAddEvent(data);
      setSlot(null);
    },
    [onAddEvent],
  );

  const handleCloseEventFormModal = useCallback(() => {
    onToggleEventFormModal();
    setSlot(null);
  }, [onToggleEventFormModal]);

  const handleSelectEvent = useCallback(
    (event: Event) => {
      setSelectedEvent(event);

      // Open the Event Details modal
      onToggleEventDetailsModal();
    },
    [onToggleEventDetailsModal],
  );

  const handleOpenConfirmModal = useCallback(() => {
    onToggleEventDetailsModal();
    onToggleConfirmModal();
  }, [onToggleConfirmModal, onToggleEventDetailsModal]);

  const handleDeleteEvent = useCallback(() => {
    onDeleteEvent(selectedEventId);

    // Close the Confirm Delete modal
    onToggleConfirmModal();
  }, [onDeleteEvent, selectedEventId, onToggleConfirmModal]);

  const eventDetailsModalsHeader = useMemo(
    () => (
      <Flex flex={1} justifyContent="space-between" alignItems="center">
        <Heading fontWeight="semibold">Event Details</Heading>
        <Flex flex={1} justifyContent="flex-end" alignItems="center" gap={3}>
          <EditIcon
            color={septenary}
            w={5}
            h={5}
            onClick={handleOpenEventFormModal}
            style={{ cursor: 'pointer' }}
            data-testid="edit-icon"
          />
          <DeleteIcon
            color={septenary}
            w={5}
            h={5}
            onClick={handleOpenConfirmModal}
            style={{ cursor: 'pointer' }}
            data-testid="delete-icon"
          />
          <CloseButton
            color={septenary}
            size="lg"
            onClick={onToggleEventDetailsModal}
            style={{ cursor: 'pointer' }}
            data-testid="close-icon"
          />
        </Flex>
      </Flex>
    ),
    [
      septenary,
      handleOpenEventFormModal,
      handleOpenConfirmModal,
      onToggleEventDetailsModal,
    ],
  );

  return (
    <Indicator isOpen={isLoading}>
      <BigCalendar
        {...rest}
        localizer={localizer}
        events={events}
        startAccessor="start"
        endAccessor="end"
        date={date}
        onNavigate={onSetDate}
        defaultView={Views.MONTH}
        view={view}
        views={[Views.MONTH, Views.WEEK, Views.DAY]}
        onView={setView}
        onSelectSlot={handleSelectSlot}
        onSelectEvent={handleSelectEvent}
        components={{ toolbar: CustomToolBar }}
        selectable
      />
      {isOpenEventFormModal && (
        <Modal
          isOpen={isOpenEventFormModal}
          onClose={handleCloseEventFormModal}
          title={`${slot ? 'Add' : 'Update'} Event`}
          body={
            <EventForm
              {...(!slot && {
                id: selectedEventId,
                eventName: selectedEventTitle,
              })}
              onCancel={handleCloseEventFormModal}
              date={eventDate}
              startTime={eventStartTime}
              endTime={eventEndTime}
              onAddEvent={handleAddEvent}
              onEditEvent={onEditEvent}
            />
          }
          haveCloseButton
        />
      )}

      {isOpenEventDetailModal && (
        <Modal
          isOpen={isOpenEventDetailModal}
          onClose={onToggleEventDetailsModal}
          title={eventDetailsModalsHeader}
          body={
            <EventDetails title={selectedEventTitle} time={selectedEventTime} />
          }
          haveCloseButton={false}
        />
      )}

      {isOpenConfirmModal && (
        <Modal
          isOpen={isOpenConfirmModal}
          onClose={onToggleConfirmModal}
          title="Delete Event"
          body={
            <ConfirmDeleteModalBody
              itemName={selectedEventTitle}
              onDelete={handleDeleteEvent}
              onCancel={onToggleConfirmModal}
            />
          }
          haveCloseButton
        />
      )}
    </Indicator>
  );
};

const CalendarMemorized = memo(Calendar, isEqual);

export default CalendarMemorized;
