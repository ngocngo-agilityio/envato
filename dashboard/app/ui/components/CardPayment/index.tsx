'use client';

import { memo, useCallback } from 'react';
import { Box, Heading, useDisclosure, useToast } from '@chakra-ui/react';
import { SubmitHandler } from 'react-hook-form';
import { AxiosError } from 'axios';

// Components
import { PinCodeModal } from '@/ui/components';
import CardBalance from './CardBalance';
import UserSelector from './UserSelector';
import EnterMoney from './EnterMoney';

// Constants
import { ERROR_MESSAGES, STATUS, SUCCESS_MESSAGES } from '@/lib/constants';

// Hooks
import {
  useAuth,
  useForm,
  useGetUserDetails,
  useMoney,
  usePinCode,
  useWallet,
} from '@/lib/hooks';

// Stores
import { authStore } from '@/lib/stores';

// Utils
import {
  customToast,
  getErrorMessageFromAxiosError,
  removeAmountFormat,
  isEnableSubmitButton,
} from '@/lib/utils';

// Types
import { TPinCodeForm, TSendMoney, TMoneyResponse } from '@/lib/interfaces';

export type TTransfer = {
  amount: string;
  memberId: string;
  userId: string;
};

const REQUIRE_FIELDS = ['amount', 'memberId'];

const CardPayment = (): JSX.Element => {
  const user = authStore((state) => state.user);

  const { setUser } = useAuth();

  const toast = useToast();

  const {
    control,
    handleSubmit: handleSubmitSendMoney,
    formState: { isSubmitting, dirtyFields },
    reset: resetSendMoneyForm,
  } = useForm<TTransfer>({
    defaultValues: {
      memberId: '',
      amount: '',
      userId: user?.id,
    },
  });

  const dirtyItems = Object.keys(dirtyFields).filter(
    (key) => dirtyFields[key as keyof TTransfer],
  );
  const shouldEnable = isEnableSubmitButton(REQUIRE_FIELDS, dirtyItems);

  const { currentWalletMoney } = useWallet(user?.id);

  const { filterDataUser } = useGetUserDetails(user?.id || '');

  const {
    isOpen: isSetPinCodeModalOpen,
    onClose: onCloseSetPinCodeModal,
    onOpen: onOpenSetPinCodeModal,
  } = useDisclosure();

  const {
    isOpen: isConfirmPinCodeModalOpen,
    onClose: onCloseConfirmPinCodeModal,
    onOpen: onOpenConfirmPinCodeModal,
  } = useDisclosure();

  const { handleSetPinCode, handleConfirmPinCode } = usePinCode();

  const {
    control: setPinCodeControl,
    handleSubmit: handleSubmitSetPinCode,
    formState: { isValid: isSetValid, isSubmitting: isSetSubmitting },
    reset: resetSetPinCodeForm,
  } = useForm<TPinCodeForm>({});

  const {
    control: confirmPinCodeControl,
    handleSubmit: handleSubmitConfirmPinCode,
    formState: { isValid: isConfirmValid, isSubmitting: isConfirmSubmitting },
    reset: resetConfirmPinCodeForm,
  } = useForm<TPinCodeForm>({
    defaultValues: {
      pinCode: '',
    },
    mode: 'onSubmit',
    reValidateMode: 'onSubmit',
  });

  const { sendMoneyToUserWallet } = useMoney();

  const getMemberId = useCallback(
    (email: string): string =>
      filterDataUser?.find(
        (user) =>
          user.email.trim().toLocaleLowerCase() === email.trim().toLowerCase(),
      )?._id || '',
    [filterDataUser],
  );

  const hasPinCode = user?.pinCode;

  const handleTransferMoneySuccess = useCallback(
    (success: { title: string; description: string }) => {
      toast(customToast(success.title, success.description, STATUS.SUCCESS));
      if (user?.bonusTimes) {
        setUser({
          user: {
            ...user,
            bonusTimes: --user.bonusTimes,
          },
        });
      }
    },
    [setUser, toast, user],
  );

  const handleTransferMoneyError = useCallback(
    (
      error: Error,
      defaultError: {
        title: string;
        description: string;
      },
    ) => {
      const responseErrorMessage = getErrorMessageFromAxiosError(
        error as AxiosError<TMoneyResponse>,
        defaultError.description,
      );

      toast(
        customToast(defaultError.title, responseErrorMessage, STATUS.ERROR),
      );
    },
    [toast],
  );

  const handleOnSubmitSendMoney = useCallback(() => {
    hasPinCode ? onOpenConfirmPinCodeModal() : onOpenSetPinCodeModal();
  }, [hasPinCode, onOpenConfirmPinCodeModal, onOpenSetPinCodeModal]);

  const onSubmitSendMoney: SubmitHandler<TTransfer> = useCallback(
    (data) => {
      const submitData: TSendMoney = {
        ...data,
        memberId: getMemberId(data.memberId),
        amount: removeAmountFormat(data.amount),
      };

      sendMoneyToUserWallet(submitData, {
        onSuccess: () =>
          handleTransferMoneySuccess(SUCCESS_MESSAGES.SEND_MONEY),
        onError: (error) =>
          handleTransferMoneyError(error, ERROR_MESSAGES.ADD_MONEY),
      });
      resetSendMoneyForm();
    },
    [
      getMemberId,
      handleTransferMoneyError,
      handleTransferMoneySuccess,
      resetSendMoneyForm,
      sendMoneyToUserWallet,
    ],
  );

  const onSubmitPinCode: SubmitHandler<TPinCodeForm> = useCallback(
    async (data) => {
      if (user) {
        data.userId = user.id;
        if (!hasPinCode) {
          try {
            await handleSetPinCode(data);

            setUser({ user: { ...user, pinCode: data.pinCode } });

            onCloseSetPinCodeModal();

            resetSetPinCodeForm();

            toast(
              customToast(
                SUCCESS_MESSAGES.SET_PIN_CODE.title,
                SUCCESS_MESSAGES.SET_PIN_CODE.description,
                STATUS.SUCCESS,
              ),
            );
          } catch (error) {
            toast(
              customToast(
                ERROR_MESSAGES.SET_PIN_CODE.title,
                ERROR_MESSAGES.SET_PIN_CODE.description,
                STATUS.ERROR,
              ),
            );
          }
        } else {
          try {
            await handleConfirmPinCode(data);
            onCloseConfirmPinCodeModal();
            resetConfirmPinCodeForm({
              pinCode: '',
            });

            await handleSubmitSendMoney(onSubmitSendMoney)();
            resetSendMoneyForm();

            toast(
              customToast(
                SUCCESS_MESSAGES.CONFIRM_PIN_CODE.title,
                SUCCESS_MESSAGES.CONFIRM_PIN_CODE.description,
                STATUS.SUCCESS,
              ),
            );
          } catch (error) {
            toast(
              customToast(
                ERROR_MESSAGES.CONFIRM_PIN_CODE.title,
                ERROR_MESSAGES.CONFIRM_PIN_CODE.description,
                STATUS.ERROR,
              ),
            );
            resetConfirmPinCodeForm();
          }
        }
      }
    },
    [
      handleConfirmPinCode,
      handleSetPinCode,
      handleSubmitSendMoney,
      hasPinCode,
      onCloseConfirmPinCodeModal,
      onCloseSetPinCodeModal,
      onSubmitSendMoney,
      resetConfirmPinCodeForm,
      resetSendMoneyForm,
      resetSetPinCodeForm,
      setUser,
      toast,
      user,
    ],
  );

  const handleCloseSetPinCodeModal = useCallback(() => {
    onCloseSetPinCodeModal();
    resetSetPinCodeForm();
  }, [onCloseSetPinCodeModal, resetSetPinCodeForm]);

  const handleCloseConfirmPinCodeModal = useCallback(() => {
    onCloseConfirmPinCodeModal();
    resetConfirmPinCodeForm();
  }, [onCloseConfirmPinCodeModal, resetConfirmPinCodeForm]);

  return (
    <>
      <Box
        p={4}
        w="full"
        bg="background.body.quaternary"
        py={{ base: 4, md: 5 }}
        px={{ base: 4, md: 10 }}
        borderRadius="lg"
      >
        <Heading
          as="h3"
          fontWeight="bold"
          color="text.primary"
          fontSize="lg"
          mb={3}
          textTransform="capitalize"
        >
          my wallet
        </Heading>

        <CardBalance balance={currentWalletMoney?.balance || 0} />

        <Box
          as="form"
          mt={4}
          onSubmit={handleSubmitSendMoney(handleOnSubmitSendMoney)}
        >
          <UserSelector control={control} listUser={filterDataUser} />
          <EnterMoney
            isDisabled={!shouldEnable || isSubmitting}
            control={control}
          />
        </Box>
      </Box>
      {/*Set/Confirm PIN code Modal */}
      <PinCodeModal
        title={
          isSetPinCodeModalOpen
            ? 'Please set the PIN code to your account'
            : 'Please enter your PIN code'
        }
        control={hasPinCode ? confirmPinCodeControl : setPinCodeControl}
        isOpen={isSetPinCodeModalOpen || isConfirmPinCodeModalOpen}
        isDisabled={
          hasPinCode
            ? !isConfirmValid || isConfirmSubmitting
            : !isSetValid || isSetSubmitting
        }
        isLoading={hasPinCode ? isConfirmSubmitting : isSetSubmitting}
        onclose={
          isSetPinCodeModalOpen
            ? handleCloseSetPinCodeModal
            : handleCloseConfirmPinCodeModal
        }
        onSubmit={
          hasPinCode
            ? handleSubmitConfirmPinCode(onSubmitPinCode)
            : handleSubmitSetPinCode(onSubmitPinCode)
        }
      />
    </>
  );
};

const CardPaymentMemorized = memo(CardPayment);

export default CardPaymentMemorized;
