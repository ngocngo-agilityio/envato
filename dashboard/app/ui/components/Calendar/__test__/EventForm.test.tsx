// Constants
import { ERROR_MESSAGES } from '@/lib/constants';

// Mocks
import {
  MOCK_CALENDAR_NOW_DATE,
  MOCK_EVENT_FORM_DATA,
  MOCK_UPDATE_EVENT_FORM,
} from '@/lib/mocks';

// Components
import { EventForm } from '@/ui/components';

const mockProps = {
  ...MOCK_EVENT_FORM_DATA,
  onCancel: jest.fn(),
  onAddEvent: jest.fn(),
  onEditEvent: jest.fn(),
};

const { render } = testLibReactUtils;

describe('EventForm component', () => {
  beforeEach(() => {
    jest.useFakeTimers({ now: MOCK_CALENDAR_NOW_DATE });
  });

  afterEach(() => {
    jest.clearAllMocks();
    jest.useRealTimers();
  });

  it('EventForm component renders correctly', () => {
    const { container } = render(<EventForm {...mockProps} />);

    expect(container).toMatchSnapshot();
  });

  it('should show the required error messages correctly', async () => {
    const { getByText, getByRole, getByLabelText } = render(
      <EventForm onCancel={mockProps.onCancel} />,
    );

    fireEvent.change(getByLabelText('Title'), {
      target: { value: MOCK_EVENT_FORM_DATA.title },
    });
    fireEvent.click(getByRole('button', { name: 'Save' }));

    await waitFor(() => {
      expect(
        getByText(ERROR_MESSAGES.FIELD_REQUIRED('Date')),
      ).toBeInTheDocument();

      expect(
        getByText(ERROR_MESSAGES.FIELD_REQUIRED('Start time')),
      ).toBeInTheDocument();

      expect(
        getByText(ERROR_MESSAGES.FIELD_REQUIRED('End time')),
      ).toBeInTheDocument();
    });
  });

  it('should show error message if the end time is less than the start time', async () => {
    const { getByText, getByRole, getByLabelText } = render(
      <EventForm {...mockProps} endTime="08:00" />,
    );

    fireEvent.change(getByLabelText('Title'), {
      target: { value: MOCK_UPDATE_EVENT_FORM.title },
    });
    fireEvent.click(getByRole('button', { name: 'Save' }));

    await waitFor(() => {
      expect(getByText(ERROR_MESSAGES.END_TIME_EVENT)).toBeInTheDocument();
    });
  });

  it('should call onAddEvent when submit form without id ', async () => {
    const { getByRole, getByLabelText } = render(<EventForm {...mockProps} />);

    fireEvent.change(getByLabelText('Date'), {
      target: { value: MOCK_UPDATE_EVENT_FORM.date },
    });
    fireEvent.click(getByRole('button', { name: 'Save' }));

    await waitFor(() => {
      expect(mockProps.onAddEvent).toHaveBeenCalled();
    });
  });

  it('should call onEditEvent when submit form with id ', async () => {
    const { getByRole, getByLabelText } = render(
      <EventForm {...mockProps} id="1" />,
    );

    fireEvent.change(getByLabelText('Date'), {
      target: { value: MOCK_UPDATE_EVENT_FORM.date },
    });
    fireEvent.click(getByRole('button', { name: 'Save' }));

    await waitFor(() => {
      expect(mockProps.onEditEvent).toHaveBeenCalled();
    });
  });
});
