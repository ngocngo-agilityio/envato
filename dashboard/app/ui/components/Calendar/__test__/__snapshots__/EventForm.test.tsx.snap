// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EventForm component EventForm component renders correctly 1`] = `
<div>
  <form
    class="chakra-stack css-tl3ftk"
  >
    <div
      class="css-15kzxzi"
    >
      <div
        class="chakra-form-control css-1kxonj9"
        role="group"
      >
        <label
          class="chakra-form__label css-1rh809e"
          for="field-:r0:"
          id="field-:r0:-label"
        >
          Title
        </label>
        <div
          class="chakra-input__group css-1091msq"
          data-group="true"
        >
          <input
            class="chakra-input css-1ji05xh"
            id="field-:r0:"
            name="eventName"
            type="text"
            value=""
          />
        </div>
      </div>
    </div>
    <div
      class="css-j8td0s"
    >
      <div
        class="chakra-form-control css-1kxonj9"
        role="group"
      >
        <label
          class="chakra-form__label css-1rh809e"
          for="field-:r1:"
          id="field-:r1:-label"
        >
          Date
        </label>
        <div
          class="chakra-input__group css-1091msq"
          data-group="true"
        >
          <input
            class="chakra-input css-1ji05xh"
            id="field-:r1:"
            name="date"
            type="date"
            value="2024-05-27"
          />
        </div>
      </div>
    </div>
    <div
      class="css-bwbm8i"
    >
      <div
        class="css-j8td0s"
      >
        <div
          class="chakra-form-control css-1kxonj9"
          role="group"
        >
          <label
            class="chakra-form__label css-1rh809e"
            for="field-:r2:"
            id="field-:r2:-label"
          >
            Start time
          </label>
          <div
            class="chakra-input__group css-1091msq"
            data-group="true"
          >
            <input
              class="chakra-input css-1127nk9"
              id="field-:r2:"
              name="startTime"
              type="time"
              value="10:00"
            />
          </div>
        </div>
      </div>
      <div
        class="css-j8td0s"
      >
        <div
          class="chakra-form-control css-1kxonj9"
          role="group"
        >
          <label
            class="chakra-form__label css-1rh809e"
            for="field-:r3:"
            id="field-:r3:-label"
          >
            End time
          </label>
          <div
            class="chakra-input__group css-1091msq"
            data-group="true"
          >
            <input
              class="chakra-input css-1ji05xh"
              id="field-:r3:"
              name="endTime"
              type="time"
              value="11:00"
            />
          </div>
        </div>
      </div>
    </div>
    <div
      class="css-6dra2j"
    >
      <button
        class="chakra-button css-1il83h3"
        disabled=""
        type="submit"
      >
        Save
      </button>
      <button
        class="chakra-button css-t98rwl"
        type="button"
      >
        Cancel
      </button>
    </div>
  </form>
  <span
    hidden=""
    id="__chakra_env"
  />
</div>
`;
